dlt/extract/pipe.py

```diff
@@ -822,34 +682,58 @@ def _get_source_item(self) -> ResolvablePipeItem:
             # if too many new sources is added we switch to fifo not to exhaust them
             if (
                 self._next_item_mode == "fifo"
-                or (sources_count - self._initial_sources_count) >= self.max_parallel_items
+                or sources_count - self._initial_sources_count
+                >= self._worker_pool.max_parallel_items
             ):
                 self._current_source_index = sources_count - 1
             else:
                 self._current_source_index = (self._current_source_index - 1) % sources_count
             while True:
-                # if we have checked all sources once and all returned None, then we can sleep a bit
+                # if we have checked all sources once and all returned None, then lets sleep a bit or wait for a free worker slot
                 if self._current_source_index == first_evaluated_index:
-                    sleep(self.futures_poll_interval)
+                    self._worker_pool.wait_for_free_slot()
                 # get next item from the current source
+                # gen, step, pipe, meta = self._sources[self._current_source_index]
                 gen, step, pipe, meta = self._sources[self._current_source_index]
                 set_current_pipe_name(pipe.name)
-                if (item := next(gen)) is not None:
+
+                pipe_item = next(gen)
+                if pipe_item is not None:
                     # full pipe item may be returned, this is used by ForkPipe step
                     # to redirect execution of an item to another pipe
-                    if isinstance(item, ResolvablePipeItem):
-                        return item
-                    else:
+                    # else
+                    if not isinstance(pipe_item, ResolvablePipeItem):
                         # keep the item assigned step and pipe when creating resolvable item
-                        if isinstance(item, DataItemWithMeta):
-                            return ResolvablePipeItem(item.data, step, pipe, item.meta)
+                        if isinstance(pipe_item, DataItemWithMeta):
+                            pipe_item = ResolvablePipeItem(
+                                pipe_item.data, step, pipe, pipe_item.meta
+                            )
                         else:
-                            return ResolvablePipeItem(item, step, pipe, meta)
+                            pipe_item = ResolvablePipeItem(pipe_item, step, pipe, meta)
+
+                    if isinstance(pipe_item.item, Awaitable) or callable(pipe_item.item):
```

Comment:
this duplication looks really suspicious. AFAIK we were returning Awaitable/Callable item to the __next__ function and resolved the item there. I do not see a reason for this code to be there. if this is a result of the code refactor then IMO something is still not right there

Reply:
It's an optimization, but I'm not totally happy with how this looks either. Want to give it a little more love.  
The basic idea behind this is that we start multiple futures on each `__next__`  iteration, instead of just adding one future and doing the whole dance "poll -> iterate all sources -> repeat" each time. It seems to make a big difference.

Reply:
My benchmark with and without this block:

  ```
  Done in 10.724883079528809 seconds
  Extracted 5000 items total from 5 resources

  Done in 20.511144161224365 seconds
  Extracted 5000 items total from 5 resources
  ```